source/extensions/io_socket/io_uring/io_handle_impl.cc

```diff
@@ -0,0 +1,357 @@
+                                                 const IoUringFactory& io_uring_factory, os_fd_t fd,
+                                                 bool socket_v6only, absl::optional<int> domain)
+    : read_buffer_size_(read_buffer_size), io_uring_factory_(io_uring_factory), fd_(fd),
+      socket_v6only_(socket_v6only), domain_(domain) {}
+
+IoUringSocketHandleImpl::~IoUringSocketHandleImpl() {
+  if (SOCKET_VALID(fd_)) {
+    IoUringSocketHandleImpl::close();
+  }
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::close() {
+  ASSERT(SOCKET_VALID(fd_));
+  io_uring_factory_.getOrCreateUring().prepareClose(fd_);
+  if (isLeader()) {
+    io_uring_factory_.getOrCreateUring().unregisterEventfd();
+    file_event_adapter_.reset();
+  }
+  SET_SOCKET_INVALID(fd_);
+  return Api::ioCallUint64ResultNoError();
+}
+
+bool IoUringSocketHandleImpl::isOpen() const { return SOCKET_VALID(fd_); }
+Api::IoCallUint64Result IoUringSocketHandleImpl::readv(uint64_t /* max_length */,
+                                                       Buffer::RawSlice* /* slices */,
+                                                       uint64_t /* num_slice */) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                      absl::optional<uint64_t> max_length_opt) {
+  const uint64_t max_length = max_length_opt.value_or(UINT64_MAX);
+  if (max_length == 0) {
+    return Api::ioCallUint64ResultNoError();
+  }
+
+  if (bytes_to_read_ == 0) {
+    return Api::IoCallUint64Result(
+        0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                           Network::IoSocketError::deleteIoError));
+  }
+
+  ASSERT(read_buf_ != nullptr);
+  auto fragment = new Buffer::BufferFragmentImpl(
+      read_buf_.get(), bytes_to_read_,
+      [](const void* /*data*/, size_t /*len*/, const Buffer::BufferFragmentImpl* this_fragment) {
+        delete this_fragment;
+      });
+  buffer.addBufferFragment(*fragment);
+  is_read_added_ = false;
+
+  uint64_t len = bytes_to_read_;
+  bytes_to_read_ = 0;
+  return Api::IoCallUint64Result(len,
+                                 Api::IoErrorPtr(nullptr, Network::IoSocketError::deleteIoError));
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::writev(const Buffer::RawSlice* /*slices */,
+                                                        uint64_t /*num_slice*/) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::IoCallUint64Result IoUringSocketHandleImpl::write(Buffer::Instance& buffer) {
+  auto length = buffer.length();
+  ASSERT(length > 0);
+  std::list<Buffer::SliceDataPtr> slice_data;
+  while (buffer.length() > 0) {
+    Buffer::SliceDataPtr data = buffer.extractMutableFrontSlice();
+    slice_data.push_back(std::move(data));
+  }
+  io_uring_factory_.getOrCreateUring().prepareWrite(fd_, std::move(slice_data));
```

Comment:
It doesn't seem to be possible for prepareWrite to provide any back pressure.  Is there an API that can be used to query how many more slices it is reasonable for the io_uring to accept and avoid infinite buffering in the kernel?

Reply:
The slices are not buffered in the kernel, but end up referenced in the submission queue entries (the queue is mmap'ed into Envoy's address space from the kernel). If the queue is full (= `io_uring_size`) and a new entry fails to be added, I submit the already queued entries to the kernel and try again adding the new entry. I'm not sure it's the right way though.